File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-8.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func goodNodes(root *TreeNode) int {
    var good int

	var dfs func(*TreeNode, int)
	dfs = func(root *TreeNode, maxN int) {
		if root == nil {
			return
		}
		if root.Val >= maxN {
			good++
		}
		maxN = max(maxN, root.Val)
		dfs(root.Left, maxN)
		dfs(root.Right, maxN)
	}

	dfs(root, -101)

	return good
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-8.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
     /**
     * @param {TreeNode} root1
     * @param {TreeNode} root2
     * @return {boolean}
     */
    dfs(root1, root2){
        if (!root1 && !root2) {return true}
        if (!root1 || !root2) {return false}

        if (root1.val === root2.val) {
            return this.dfs(root1.left, root2.left) && this.dfs(root1.right, root2.right)
        }
    }
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        if (!root) {return false}
        if (!subRoot) {return true}

        if (this.dfs(root, subRoot)) {return true}

        return this.isSubtree(root.left, subRoot) || this.isSubtree(root.right, subRoot)
    }
}
